iOS: Text insertion caret can get out of sync with the page content

hortont424 opened the pull request:
#### 31c6ba7d4ea89e6add0f115823031e728f092663
<pre>
iOS: Text insertion caret can get out of sync with the page content
<a href="https://bugs.webkit.org/show_bug.cgi?id=244105">https://bugs.webkit.org/show_bug.cgi?id=244105</a>
rdar://96977291

Reviewed by Wenson Hsieh.

When remote layer tree transactions are slow to flush, intermediate EditorState
updates (from e.g. interpretKeyEvent or other sources) can cause the
UI-process-drawn text insertion caret to jump ahead of the current page tiles.
This is especially egregious when deleting a character, as the insertion
point shows up in between the soon-to-be-deleted character and the previous
character *before* the deleted one disappears.

The problem is exacerbated on slow hardware, or expensive pages.

To fix this, separate &quot;visual&quot; parts of EditorState out into their own
sub-struct, and only update it for EditorState updates that come inside
remote layer tree transactions. Other changes to EditorState will propagate
forward the existing &quot;visual&quot; data, until a new transaction comes in to replace it.

* Source/WebKit/Shared/EditorState.cpp:
(WebKit::EditorState::encode const):
(WebKit::EditorState::decode):
(WebKit::EditorState::PostLayoutData::encode const):
(WebKit::EditorState::PostLayoutData::decode):
(WebKit::EditorState::VisualData::encode const):
(WebKit::EditorState::VisualData::decode):
(WebKit::operator&lt;&lt;):
* Source/WebKit/Shared/EditorState.h:
(WebKit::EditorState::hasPostLayoutData const):
(WebKit::EditorState::ensurePostLayoutData):
(WebKit::EditorState::hasVisualData const):
(WebKit::EditorState::hasPostLayoutAndVisualData const):
(WebKit::EditorState::postLayoutData):
(WebKit::EditorState::postLayoutData const const):
(WebKit::EditorState::visualData):
(WebKit::EditorState::visualData const const):
(WebKit::EditorState::ensureVisualData):
Factor out some geometry into VisualData.
Add hasPostLayoutData and make isMissingPostLayoutData private, to make
the API the same between postLayoutData and visualData.

* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
(WebKit::RemoteLayerTreeDrawingAreaProxy::commitLayerTree):
Update EditorState, explicitly allowing VisualData merging.

* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::updateFontAttributesAfterEditorStateChange):
(WebKit::WebPageProxy::updateEditorState):
* Source/WebKit/UIProcess/WebPageProxy.h:
For UI-side compositing clients, make updateEditorState only merge VisualData
changes if they come in a layer tree transaction. Otherwise, we maintain the
existing VisualData. This gets a little tricky, because we need to merge
VisualData changes even if the new EditorState is otherwise stale, and
also need to avoid all of this work (and just take the current incoming
EditorState) for non-UI-side-compositing clients.

* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
(WebKit::WebPage::getPlatformEditorStateCommon const):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::editorState const):
Generate VisualData whenever we would generate PostLayoutData.

(WebKit::WebPage::willCommitLayerTree):
(WebKit::WebPage::sendEditorStateUpdate):
(WebKit::WebPage::scheduleFullEditorStateUpdate):
Ensure that if we have scheduled *any* EditorState update since the last transaction,
the next layer tree transaction includes an EditorState update (so that we receive
and merge VisualData).

* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/glib/WebPageGLib.cpp:
(WebKit::WebPage::getPlatformEditorState const):
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::getPlatformEditorState const):
* Source/WebKit/WebProcess/WebPage/mac/WebPageMac.mm:
(WebKit::WebPage::getPlatformEditorState const):
* Source/WebKit/UIProcess/ios/WKContentView.mm:
(-[WKContentView _didCommitLayerTree:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(dictionaryRepresentationForEditorState):
* Source/WebKit/UIProcess/API/glib/WebKitEditorState.cpp:
(webkitEditorStateChanged):
* Source/WebKit/UIProcess/API/gtk/WebKitWebViewBase.cpp:
(webkitWebViewBaseUpdateTextInputState):
* Source/WebKit/UIProcess/API/wpe/WPEView.cpp:
(WKWPE::View::selectionDidChange):
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(WebKit::WKSelectionDrawingInfo::WKSelectionDrawingInfo):
(-[WKContentView shouldHideSelectionWhenScrolling]):
(-[WKContentView rectToRevealWhenZoomingToFocusedElement]):
(-[WKContentView _selectionClipRect]):
(-[WKContentView canShowNonEmptySelectionView]):
(-[WKContentView webSelectionRects]):
(-[WKContentView _lookupForWebView:]):
(-[WKContentView _shareForWebView:]):
(-[WKContentView _translateForWebView:]):
(-[WKContentView _addShortcutForWebView:]):
(-[WKContentView _cascadeInteractionTintColor]):
(-[WKContentView _showDictionary:]):
(-[WKContentView removeBackgroundMenu]):
(-[WKContentView doAfterComputingImageAnalysisResultsForBackgroundRemoval:]):
(-[WKContentView textFirstRect]):
(-[WKContentView textLastRect]):
(-[WKContentView textInRange:]):
(-[WKContentView selectedTextRange]):
(-[WKContentView markedTextRange]):
(-[WKContentView hasText]):
(-[WKContentView _updateSelectionAssistantSuppressionState]):
(-[WKContentView _selectionChanged]):
(-[WKContentView _updateChangedSelection:]):
* Source/WebKit/UIProcess/ios/WebPageProxyIOS.mm:
(WebKit::WebPageProxy::interpretKeyEvent):
(WebKit::WebPageProxy::dispatchDidUpdateEditorState):
(WebKit::WebPageProxy::selectionBoundingRectInRootViewCoordinates const):
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::selectionDidChange):
(WebKit::WebViewImpl::requestCandidatesForSelectionIfNeeded):
(WebKit::WebViewImpl::handleRequestedCandidates):
(WebKit::WebViewImpl::handleAcceptedCandidate):
(WebKit::WebViewImpl::updateTextTouchBar):
Adopt VisualData everywhere.

Canonical link: <a href="https://commits.webkit.org/255884@main">https://commits.webkit.org/255884@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/145ea892a6e9ecfa8655da5bff96f058d4d90d3a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/93811 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/3005 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/24404 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/103447 "Built successfully") | 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/3021 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/31247 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/86137 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/99458 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/99472 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/2154 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/80241 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/29179 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/84080 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/83796 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/72127 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/37642 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/17624 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/35501 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/18884 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4062 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/39380 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/41447 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/41314 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/38115 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

whsieh commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -1269,7 +1269,8 @@ EditorState WebPage::editorState(ShouldPerformLayout shouldPerformLayout) const
+        result.ensurePostLayoutData();
+        result.ensureVisualData();
 
         auto& postLayoutData = result.postLayoutData();
```
Nit - instead of `result.ensurePostLayoutData();` above, it might be cleaner to just do:

```
auto& postLayoutData = result.ensurePostLayoutData();
…
```

hortont424 replied:
Definitely.

whsieh commented on Source/WebKit/UIProcess/mac/WebViewImpl.mm:
```diff
@@ -3119,7 +3119,7 @@ static String commandNameForSelector(SEL selector)
         return;
 
-    if (editorState.isMissingPostLayoutData)
+    if (!editorState.hasPostLayoutAndVisualData())
```
It seems like this logic only requires up-to-date post layout data, and not visual data; should this check be `hasPostLayoutData()` instead of `hasPostLayoutAndVisualData()`?

hortont424 replied:
Heh, I went over these a few times, not surprised that I still managed to miss some.

whsieh commented on Source/WebKit/UIProcess/mac/WebViewImpl.mm:
```diff
@@ -3151,7 +3151,7 @@ static String commandNameForSelector(SEL selector)
     // FIXME: It's pretty lame that we have to depend on the most recent EditorState having post layout data,
     // and that we just bail if it is missing.
-    if (editorState.isMissingPostLayoutData)
+    if (!editorState.hasPostLayoutAndVisualData())
```
(Ditto, re: `hasPostLayoutData()` vs. `hasPostLayoutAndVisualData()`).

whsieh reviewed (approved):
r=mews

hortont424 committed:
iOS: Text insertion caret can get out of sync with the page content
https://bugs.webkit.org/show_bug.cgi?id=244105
rdar://96977291

Reviewed by Wenson Hsieh.

When remote layer tree transactions are slow to flush, intermediate EditorState
updates (from e.g. interpretKeyEvent or other sources) can cause the
UI-process-drawn text insertion caret to jump ahead of the current page tiles.
This is especially egregious when deleting a character, as the insertion
point shows up in between the soon-to-be-deleted character and the previous
character *before* the deleted one disappears.

The problem is exacerbated on slow hardware, or expensive pages.

To fix this, separate "visual" parts of EditorState out into their own
sub-struct, and only update it for EditorState updates that come inside
remote layer tree transactions. Other changes to EditorState will propagate
forward the existing "visual" data, until a new transaction comes in to replace it.

* Source/WebKit/Shared/EditorState.cpp:
(WebKit::EditorState::encode const):
(WebKit::EditorState::decode):
(WebKit::EditorState::PostLayoutData::encode const):
(WebKit::EditorState::PostLayoutData::decode):
(WebKit::EditorState::VisualData::encode const):
(WebKit::EditorState::VisualData::decode):
(WebKit::operator<<):
* Source/WebKit/Shared/EditorState.h:
(WebKit::EditorState::hasPostLayoutData const):
(WebKit::EditorState::ensurePostLayoutData):
(WebKit::EditorState::hasVisualData const):
(WebKit::EditorState::hasPostLayoutAndVisualData const):
(WebKit::EditorState::postLayoutData):
(WebKit::EditorState::postLayoutData const const):
(WebKit::EditorState::visualData):
(WebKit::EditorState::visualData const const):
(WebKit::EditorState::ensureVisualData):
Factor out some geometry into VisualData.
Add hasPostLayoutData and make isMissingPostLayoutData private, to make
the API the same between postLayoutData and visualData.

* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
(WebKit::RemoteLayerTreeDrawingAreaProxy::commitLayerTree):
Update EditorState, explicitly allowing VisualData merging.

* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::updateFontAttributesAfterEditorStateChange):
(WebKit::WebPageProxy::updateEditorState):
* Source/WebKit/UIProcess/WebPageProxy.h:
For UI-side compositing clients, make updateEditorState only merge VisualData
changes if they come in a layer tree transaction. Otherwise, we maintain the
existing VisualData. This gets a little tricky, because we need to merge
VisualData changes even if the new EditorState is otherwise stale, and
also need to avoid all of this work (and just take the current incoming
EditorState) for non-UI-side-compositing clients.

* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
(WebKit::WebPage::getPlatformEditorStateCommon const):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::editorState const):
Generate VisualData whenever we would generate PostLayoutData.

(WebKit::WebPage::willCommitLayerTree):
(WebKit::WebPage::sendEditorStateUpdate):
(WebKit::WebPage::scheduleFullEditorStateUpdate):
Ensure that if we have scheduled *any* EditorState update since the last transaction,
the next layer tree transaction includes an EditorState update (so that we receive
and merge VisualData).

* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/glib/WebPageGLib.cpp:
(WebKit::WebPage::getPlatformEditorState const):
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::getPlatformEditorState const):
* Source/WebKit/WebProcess/WebPage/mac/WebPageMac.mm:
(WebKit::WebPage::getPlatformEditorState const):
* Source/WebKit/UIProcess/ios/WKContentView.mm:
(-[WKContentView _didCommitLayerTree:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(dictionaryRepresentationForEditorState):
* Source/WebKit/UIProcess/API/glib/WebKitEditorState.cpp:
(webkitEditorStateChanged):
* Source/WebKit/UIProcess/API/gtk/WebKitWebViewBase.cpp:
(webkitWebViewBaseUpdateTextInputState):
* Source/WebKit/UIProcess/API/wpe/WPEView.cpp:
(WKWPE::View::selectionDidChange):
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(WebKit::WKSelectionDrawingInfo::WKSelectionDrawingInfo):
(-[WKContentView shouldHideSelectionWhenScrolling]):
(-[WKContentView rectToRevealWhenZoomingToFocusedElement]):
(-[WKContentView _selectionClipRect]):
(-[WKContentView canShowNonEmptySelectionView]):
(-[WKContentView webSelectionRects]):
(-[WKContentView _lookupForWebView:]):
(-[WKContentView _shareForWebView:]):
(-[WKContentView _translateForWebView:]):
(-[WKContentView _addShortcutForWebView:]):
(-[WKContentView _cascadeInteractionTintColor]):
(-[WKContentView _showDictionary:]):
(-[WKContentView removeBackgroundMenu]):
(-[WKContentView doAfterComputingImageAnalysisResultsForBackgroundRemoval:]):
(-[WKContentView textFirstRect]):
(-[WKContentView textLastRect]):
(-[WKContentView textInRange:]):
(-[WKContentView selectedTextRange]):
(-[WKContentView markedTextRange]):
(-[WKContentView hasText]):
(-[WKContentView _updateSelectionAssistantSuppressionState]):
(-[WKContentView _selectionChanged]):
(-[WKContentView _updateChangedSelection:]):
* Source/WebKit/UIProcess/ios/WebPageProxyIOS.mm:
(WebKit::WebPageProxy::interpretKeyEvent):
(WebKit::WebPageProxy::dispatchDidUpdateEditorState):
(WebKit::WebPageProxy::selectionBoundingRectInRootViewCoordinates const):
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::selectionDidChange):
(WebKit::WebViewImpl::requestCandidatesForSelectionIfNeeded):
(WebKit::WebViewImpl::handleRequestedCandidates):
(WebKit::WebViewImpl::handleAcceptedCandidate):
(WebKit::WebViewImpl::updateTextTouchBar):
Adopt VisualData everywhere.

Canonical link: https://commits.webkit.org/255884@main
